src/JSInterop/Microsoft.JSInterop/src/JSRuntimeBase.cs

```diff
@@ -20,40 +22,99 @@ public abstract class JSRuntimeBase : IJSRuntime
         private readonly ConcurrentDictionary<long, object> _pendingTasks
             = new ConcurrentDictionary<long, object>();
 
+        private readonly ConcurrentDictionary<long, CancellationTokenRegistration> _cancellationRegistrations =
+            new ConcurrentDictionary<long, CancellationTokenRegistration>();
+
         internal DotNetObjectRefManager ObjectRefManager { get; } = new DotNetObjectRefManager();
 
+        /// <summary>
+        /// Gets or sets the default timeout for asynchronous JavaScript calls.
+        /// </summary>
+        protected TimeSpan? DefaultAsyncCallTimeout { get; set; }
+
         /// <summary>
         /// Invokes the specified JavaScript function asynchronously.
         /// </summary>
         /// <typeparam name="T">The JSON-serializable return type.</typeparam>
         /// <param name="identifier">An identifier for the function to invoke. For example, the value <code>"someScope.someFunction"</code> will invoke the function <code>window.someScope.someFunction</code>.</param>
         /// <param name="args">JSON-serializable arguments.</param>
+        /// <param name="cancellationToken">A cancellation token to signal the cancellation of the operation.</param>
         /// <returns>An instance of <typeparamref name="T"/> obtained by JSON-deserializing the return value.</returns>
-        public Task<T> InvokeAsync<T>(string identifier, params object[] args)
+        public Task<T> InvokeAsync<T>(string identifier, IEnumerable<object> args, CancellationToken cancellationToken = default)
         {
-            // We might consider also adding a default timeout here in case we don't want to
-            // risk a memory leak in the scenario where the JS-side code is failing to complete
-            // the operation.
-
             var taskId = Interlocked.Increment(ref _nextPendingTaskId);
-            var tcs = new TaskCompletionSource<T>();
+            var tcs = new TaskCompletionSource<T>(TaskContinuationOptions.RunContinuationsAsynchronously);
+            if (cancellationToken != default)
+            {
+                _cancellationRegistrations[taskId] = cancellationToken.Register(() =>
+                {
+                    tcs.TrySetCanceled();
```

Comment:
Should you pass in the `CancellationToken` here?

Reply:
That's the point of that? Just curious cause I can't figure out from the documentation.

Reply:
I think it's just to avoid having to capture `tcs` in the closure, which costs an allocation.

In this case it doesn't help us avoid an allocation, since we also have to capture `taskId` anyway.

Reply:
> I think it's just to avoid having to capture `tcs` in the closure, which costs an allocation.
> 
> In this case it doesn't help us avoid an allocation, since we also have to capture `taskId` anyway.

I think pranav is referring to passing the cancellation token to trySetCancelled, which I'm not sure how would avoid the allocation as that would require receiving it as a parameter to Register

Reply:
I haven't been able (looking at the corefx source code) to determine why passing the cancellation token is important, as AFAIK it doesn't get reflected publicly in any place. (It ends up on an internal Task.CancellationToken property)

Reply:
That said, I added it because its uber cheap